satpy/readers/safe_sar_c.py

```diff
@@ -120,47 +122,118 @@ def __init__(self, filename, filename_info, filetype_info, calfh, noisefh):
+            logger.debug('Reading calibration data.')
 
-        cal = self.calibration.get_calibration('gamma', data.shape)
-        cal_constant = self.calibration.get_calibration_constant()
+            cal = self.calibration.get_calibration('gamma', data.shape)
+            cal_constant = self.calibration.get_calibration_constant()
 
-        logger.debug('Calibrating.')
+            logger.debug('Calibrating.')
 
-        # val = np.sqrt((data ** 2. + cal_constant - noise) / sigma ** 2)
-        # val = np.sqrt((data ** 2. - noise))
-        # data = (data ** 2. + cal_constant - noise) / sigma_sqr
-        data **= 2
-        data += cal_constant - noise
-        data /= cal
-        data[data < 0] = 0
-        del noise, cal
-        proj = Projectable(np.sqrt(data),
-                           copy=False,
-                           units='sigma')
+            # val = np.sqrt((data ** 2. + cal_constant - noise) / sigma ** 2)
+            # val = np.sqrt((data ** 2. - noise))
+            # data = (data ** 2. + cal_constant - noise) / sigma_sqr
+            data **= 2
+            data += cal_constant - noise
+            data /= cal
+            data[data < 0] = 0
+            del noise, cal
+
+            proj = Projectable(np.sqrt(data),
+                               copy=False,
+                               units='sigma')
+        del band
         return proj
 
+    def get_lonlats(self, band, array_shape):
+        """
+        Obtain GCPs and construct latitude and longitude arrays
+ 	    Args:
+           band (gdal band): Measurement band which comes with GCP's
+           array_shape (tuple) : The size of the data array
+        Returns:
+           coordinates (tuple): A tuple with longitude and latitude arrays
+        """
+        (xpoints, ypoints), (gcp_lons, gcp_lats), (fine_cols, fine_rows) = self.get_gcps(band, array_shape)
+        satint = GeoInterpolator((gcp_lons, gcp_lats),
+		                         (ypoints, xpoints),
+		                         (fine_rows, fine_cols), 2,2)
+
+        longitudes, latitudes = satint.interpolate()
+
+        # FIXME: check if the array is C-contigious, and make it such if it isn't
+        if longitudes.flags['CONTIGUOUS'] is False:
+            longitudes = np.ascontiguousarray(longitudes)
+        if latitudes.flags['CONTIGUOUS'] is False:
+            latitudes = np.ascontiguousarray(latitudes)
+
+        return longitudes, latitudes
+
+    def get_gcps(self, band, array_shape):
+        """
+        Read GCP from the GDAL band
+
+        Args:
+           band (gdal band): Measurement band which comes with GCP's
+           coordinates (tuple): A tuple with longitude and latitude arrays
+
+        Returns:
+           points (tuple): Pixel and Line indices 1d arrays
+           gcp_coords (tuple): longitude and latitude 1d arrays
+           fine_grid_points (tuple): pixel and line indices 1d arrays for the new coordinate grid
```

Comment:
It looks strange to have the fine grid computation here, since it has nothing to do with GCPs